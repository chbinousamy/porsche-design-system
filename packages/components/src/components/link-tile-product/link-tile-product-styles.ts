import { getCss, Theme } from '../../utils';
import {
  addImportantToEachRule,
  colorSchemeStyles,
  getThemedColors,
  getTransition,
  hostHiddenStyles,
  hoverMediaQuery,
  prefersColorSchemeDarkMediaQuery,
} from '../../styles';
import {
  borderRadiusLarge,
  borderRadiusMedium,
  fontLineHeight,
  fontWeightRegular,
  fontWeightSemiBold,
  spacingFluidMedium,
  spacingFluidSmall,
  spacingFluidXSmall,
  textSmallStyle,
  textXSmallStyle,
  textXXSmallStyle,
} from '@porsche-design-system/utilities-v2';

export const getComponentCss = (theme: Theme): string => {
  const { primaryColor, contrastHighColor, backgroundSurfaceColor } = getThemedColors(theme);
  const {
    primaryColor: primaryColorDark,
    contrastHighColor: contrastHighColorDark,
    backgroundSurfaceColor: backgroundSurfaceColorDark,
  } = getThemedColors('dark');
  return getCss({
    '@global': {
      ':host': {
        position: 'relative',
        ...addImportantToEachRule({
          ...colorSchemeStyles,
          ...hostHiddenStyles,
        }),
      },
      ...addImportantToEachRule({
        '::slotted': {
          '&(picture)': {
            display: 'block',
            overflow: 'hidden',
            aspectRatio: '8 / 9',
            objectFit: 'cover',
            width: '100%',
            borderRadius: borderRadiusLarge,
          },
          '&(img)': {
            aspectRatio: '8 / 9',
            objectFit: 'cover',
            width: '100%',
            borderRadius: borderRadiusLarge,
          },
        },
      }),
    },
    root: {
      display: 'block',
      textDecoration: 'none',
      borderRadius: borderRadiusMedium,
      padding: spacingFluidSmall,
      color: primaryColor,
      backgroundColor: backgroundSurfaceColor,
      ...prefersColorSchemeDarkMediaQuery(theme, {
        color: primaryColorDark,
        backgroundColor: backgroundSurfaceColorDark,
      }),
      ...hoverMediaQuery({
        '&:hover .image-container': {
          transform: 'scale3d(1.05,1.05,1.05)',
        },
      }),
    },
    header: {
      display: 'flex',
      overflow: 'hidden',
      paddingRight: fontLineHeight, // Reserve space for like button
    },
    'like-button': {
      position: 'absolute',
      top: spacingFluidSmall,
      right: spacingFluidSmall,
      height: 'fit-content',
      ...hoverMediaQuery({
        '&:hover': {
          cursor: 'pointer',
        },
      }),
    },
    text: {
      textAlign: 'center',
      margin: 0,
      '&__heading-container': {
        height: `calc(${fontLineHeight} * 2)`,
        display: 'flex',
        alignItems: 'flex-end',
        justifyContent: 'center',
        overflow: 'hidden',
      },
      '&__heading': {
        ...textSmallStyle,
        fontWeight: fontWeightSemiBold,
        maxHeight: `calc(${fontLineHeight} * 2)`,
        display: '-webkit-box',
        WebkitLineClamp: 2,
        WebkitBoxOrient: 'vertical',
        overflow: 'hidden',
        textOverflow: 'ellipsis',
      },
      '&__price': {
        ...textXSmallStyle,
        fontWeight: fontWeightRegular,
      },
      '&__info': {
        ...textXXSmallStyle,
        fontWeight: fontWeightRegular,
        paddingBottom: spacingFluidXSmall,
        color: contrastHighColor,
        ...prefersColorSchemeDarkMediaQuery(theme, {
          color: contrastHighColorDark,
        }),
      },
    },
    'image-container': {
      display: 'block',
      aspectRatio: '8 / 9',
      margin: `${spacingFluidSmall} ${spacingFluidMedium} ${spacingFluidXSmall} ${spacingFluidMedium} `,
      borderRadius: borderRadiusLarge,
      transition: getTransition('transform', 'moderate'),
    },
  });
};
