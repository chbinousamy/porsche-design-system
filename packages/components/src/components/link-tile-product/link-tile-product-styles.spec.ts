import { getComponentCss } from './link-tile-product-styles';

describe('getComponentCss()', () => {
  it.each<Parameters<typeof getComponentCss>>([
    [true, true, '3:4', 'light'],
    [true, false, '3:4', 'dark'],
    [true, false, '3:4', 'auto'],
    [false, false, '9:16', 'dark'],
    [true, false, { base: '3:4', xs: '9:16' }, 'light'],
  ])(
    'should return correct css for hasLikeButton: %s, hasSlottedAnchor: %s, aspectRatio: %j and theme: %s',
    (...args) => {
      expect(getComponentCss(...args)).toMatchSnapshot();
    }
  );
});
