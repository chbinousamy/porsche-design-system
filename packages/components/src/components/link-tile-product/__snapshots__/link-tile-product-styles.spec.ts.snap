// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getComponentCss() should return correct css for hasLikeButton: false, hasSlottedAnchor: false, aspectRatio: "9:16" and theme: dark 1`] = `
":host {
  display: block;
  position: relative;
  color-scheme: light dark !important;
}
:host([hidden]) {
  display: none !important;
}
::slotted(img), ::slotted(picture) {
  display: block !important;
  width: 100% !important;
  height: 100% !important;
  object-fit: cover !important;
  overflow: hidden !important;
  border-radius: 12px !important;
  mix-blend-mode: var(--p-link-tile-product-mix-blend-mode) !important;
}
.link-overlay {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  border-radius: 8px;
}
.link-overlay:focus {
  outline: 2px solid #1A44EA;
  outline-offset: 2px;
}
.link-overlay:focus:not(:focus-visible) {
  outline-color: transparent;
}
.root {
  display: flex;
  flex-direction: column;
  aspect-ratio: 9 / 16;
  overflow: hidden;
  box-sizing: border-box;
  border-radius: 8px;
  padding: clamp(8px, 0.5vw + 6px, 16px);
  color: #FBFCFF;
  background-color: #212225;
}
.header {
  display: flex;
  justify-content: space-between;
  font-size: clamp(0.81rem, 0.23vw + 0.77rem, 0.88rem);
}
.header slot[name="header"] {
  display: block;
}
.text-container {
  display: flex;
  justify-content: center;
  flex-direction: column;
  margin: auto;
}
.text {
  text-align: center;
  margin: 0;
}
.text__heading {
  font: normal normal 600 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  padding-bottom: 2px;
  min-height: calc(calc(6px + 2.125ex) * 2);
  display: flex;
  justify-content: center;
  align-items: flex-end;
}
.text__price, .text__info {
  font-weight: 400;
}
.text__price {
  font: normal normal 400 clamp(0.81rem, 0.23vw + 0.77rem, 0.88rem)/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
}
.text__info {
  font: normal normal 400 .75rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  color: #AFB0B3;
}
.image-container {
  margin: clamp(8px, 0.5vw + 6px, 16px) auto clamp(4px, 0.25vw + 3px, 8px) auto;
  padding: 0;
  overflow: hidden;
  aspect-ratio: 8 / 9;
  transition: transform var(--p-transition-duration, 0.4s) cubic-bezier(0.25,0.1,0.25,1);
  max-height: 100%;
}
@media(min-width:760px) {
  .image-container {
    padding: 0 clamp(16px, 1.25vw + 12px, 36px);
  }
}

@media(hover:hover) {
  .root:hover .image-container {
    transform: scale3d(1.05,1.05,1.05);
  }
}
"
`;

exports[`getComponentCss() should return correct css for hasLikeButton: true, hasSlottedAnchor: false, aspectRatio: "3:4" and theme: auto 1`] = `
":host {
  display: block;
  position: relative;
  color-scheme: light dark !important;
}
:host([hidden]) {
  display: none !important;
}
::slotted(img), ::slotted(picture) {
  display: block !important;
  width: 100% !important;
  height: 100% !important;
  object-fit: cover !important;
  overflow: hidden !important;
  border-radius: 12px !important;
  mix-blend-mode: var(--p-link-tile-product-mix-blend-mode) !important;
}
.link-overlay {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  border-radius: 8px;
}
.link-overlay:focus {
  outline: 2px solid #1A44EA;
  outline-offset: 2px;
}
.link-overlay:focus:not(:focus-visible) {
  outline-color: transparent;
}
.root {
  display: flex;
  flex-direction: column;
  aspect-ratio: 3 / 4;
  overflow: hidden;
  box-sizing: border-box;
  border-radius: 8px;
  padding: clamp(8px, 0.5vw + 6px, 16px);
  color: #010205;
  background-color: #EEEFF2;
}
.header {
  display: flex;
  justify-content: space-between;
  font-size: clamp(0.81rem, 0.23vw + 0.77rem, 0.88rem);
}
.header slot[name="header"] {
  display: block;
}
.like-button {
  height: fit-content;
  position: relative;
  z-index: 2;
}
.text-container {
  display: flex;
  justify-content: center;
  flex-direction: column;
  margin: auto;
}
.text {
  text-align: center;
  margin: 0;
}
.text__heading {
  font: normal normal 600 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  padding-bottom: 2px;
  min-height: calc(calc(6px + 2.125ex) * 2);
  display: flex;
  justify-content: center;
  align-items: flex-end;
}
.text__price, .text__info {
  font-weight: 400;
}
.text__price {
  font: normal normal 400 clamp(0.81rem, 0.23vw + 0.77rem, 0.88rem)/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
}
.text__info {
  font: normal normal 400 .75rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  color: #535457;
}
.image-container {
  margin: clamp(8px, 0.5vw + 6px, 16px) auto clamp(4px, 0.25vw + 3px, 8px) auto;
  padding: 0;
  overflow: hidden;
  aspect-ratio: 8 / 9;
  transition: transform var(--p-transition-duration, 0.4s) cubic-bezier(0.25,0.1,0.25,1);
  max-height: 100%;
}
@media(min-width:760px) {
  .image-container {
    padding: 0 clamp(16px, 1.25vw + 12px, 36px);
  }
}

@media (prefers-color-scheme: dark) {
  .root {
    color: #FBFCFF;
    background-color: #212225;
  }
  .text__info {
    color: #AFB0B3;
  }
}

@media(hover:hover) {
  .root:hover .image-container {
    transform: scale3d(1.05,1.05,1.05);
  }
  .like-button:hover {
    cursor: pointer;
  }
}
"
`;

exports[`getComponentCss() should return correct css for hasLikeButton: true, hasSlottedAnchor: false, aspectRatio: "3:4" and theme: dark 1`] = `
":host {
  display: block;
  position: relative;
  color-scheme: light dark !important;
}
:host([hidden]) {
  display: none !important;
}
::slotted(img), ::slotted(picture) {
  display: block !important;
  width: 100% !important;
  height: 100% !important;
  object-fit: cover !important;
  overflow: hidden !important;
  border-radius: 12px !important;
  mix-blend-mode: var(--p-link-tile-product-mix-blend-mode) !important;
}
.link-overlay {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  border-radius: 8px;
}
.link-overlay:focus {
  outline: 2px solid #1A44EA;
  outline-offset: 2px;
}
.link-overlay:focus:not(:focus-visible) {
  outline-color: transparent;
}
.root {
  display: flex;
  flex-direction: column;
  aspect-ratio: 3 / 4;
  overflow: hidden;
  box-sizing: border-box;
  border-radius: 8px;
  padding: clamp(8px, 0.5vw + 6px, 16px);
  color: #FBFCFF;
  background-color: #212225;
}
.header {
  display: flex;
  justify-content: space-between;
  font-size: clamp(0.81rem, 0.23vw + 0.77rem, 0.88rem);
}
.header slot[name="header"] {
  display: block;
}
.like-button {
  height: fit-content;
  position: relative;
  z-index: 2;
}
.text-container {
  display: flex;
  justify-content: center;
  flex-direction: column;
  margin: auto;
}
.text {
  text-align: center;
  margin: 0;
}
.text__heading {
  font: normal normal 600 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  padding-bottom: 2px;
  min-height: calc(calc(6px + 2.125ex) * 2);
  display: flex;
  justify-content: center;
  align-items: flex-end;
}
.text__price, .text__info {
  font-weight: 400;
}
.text__price {
  font: normal normal 400 clamp(0.81rem, 0.23vw + 0.77rem, 0.88rem)/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
}
.text__info {
  font: normal normal 400 .75rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  color: #AFB0B3;
}
.image-container {
  margin: clamp(8px, 0.5vw + 6px, 16px) auto clamp(4px, 0.25vw + 3px, 8px) auto;
  padding: 0;
  overflow: hidden;
  aspect-ratio: 8 / 9;
  transition: transform var(--p-transition-duration, 0.4s) cubic-bezier(0.25,0.1,0.25,1);
  max-height: 100%;
}
@media(min-width:760px) {
  .image-container {
    padding: 0 clamp(16px, 1.25vw + 12px, 36px);
  }
}

@media(hover:hover) {
  .root:hover .image-container {
    transform: scale3d(1.05,1.05,1.05);
  }
  .like-button:hover {
    cursor: pointer;
  }
}
"
`;

exports[`getComponentCss() should return correct css for hasLikeButton: true, hasSlottedAnchor: false, aspectRatio: {"base":"3:4","xs":"9:16"} and theme: light 1`] = `
":host {
  display: block;
  position: relative;
  color-scheme: light dark !important;
}
:host([hidden]) {
  display: none !important;
}
::slotted(img), ::slotted(picture) {
  display: block !important;
  width: 100% !important;
  height: 100% !important;
  object-fit: cover !important;
  overflow: hidden !important;
  border-radius: 12px !important;
  mix-blend-mode: var(--p-link-tile-product-mix-blend-mode) !important;
}
.link-overlay {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  border-radius: 8px;
}
.link-overlay:focus {
  outline: 2px solid #1A44EA;
  outline-offset: 2px;
}
.link-overlay:focus:not(:focus-visible) {
  outline-color: transparent;
}
.root {
  display: flex;
  flex-direction: column;
  aspect-ratio: 3 / 4;
  overflow: hidden;
  box-sizing: border-box;
  border-radius: 8px;
  padding: clamp(8px, 0.5vw + 6px, 16px);
  color: #010205;
  background-color: #EEEFF2;
}
.header {
  display: flex;
  justify-content: space-between;
  font-size: clamp(0.81rem, 0.23vw + 0.77rem, 0.88rem);
}
.header slot[name="header"] {
  display: block;
}
.like-button {
  height: fit-content;
  position: relative;
  z-index: 2;
}
.text-container {
  display: flex;
  justify-content: center;
  flex-direction: column;
  margin: auto;
}
.text {
  text-align: center;
  margin: 0;
}
.text__heading {
  font: normal normal 600 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  padding-bottom: 2px;
  min-height: calc(calc(6px + 2.125ex) * 2);
  display: flex;
  justify-content: center;
  align-items: flex-end;
}
.text__price, .text__info {
  font-weight: 400;
}
.text__price {
  font: normal normal 400 clamp(0.81rem, 0.23vw + 0.77rem, 0.88rem)/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
}
.text__info {
  font: normal normal 400 .75rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  color: #535457;
}
.image-container {
  margin: clamp(8px, 0.5vw + 6px, 16px) auto clamp(4px, 0.25vw + 3px, 8px) auto;
  padding: 0;
  overflow: hidden;
  aspect-ratio: 8 / 9;
  transition: transform var(--p-transition-duration, 0.4s) cubic-bezier(0.25,0.1,0.25,1);
  max-height: 100%;
}
@media(min-width:480px) {
  .root {
    aspect-ratio: 9 / 16;
  }
}

@media(min-width:760px) {
  .image-container {
    padding: 0 clamp(16px, 1.25vw + 12px, 36px);
  }
}

@media(hover:hover) {
  .root:hover .image-container {
    transform: scale3d(1.05,1.05,1.05);
  }
  .like-button:hover {
    cursor: pointer;
  }
}
"
`;

exports[`getComponentCss() should return correct css for hasLikeButton: true, hasSlottedAnchor: true, aspectRatio: "3:4" and theme: light 1`] = `
":host {
  display: block;
  position: relative;
  color-scheme: light dark !important;
}
:host([hidden]) {
  display: none !important;
}
::slotted(img), ::slotted(picture) {
  display: block !important;
  width: 100% !important;
  height: 100% !important;
  object-fit: cover !important;
  overflow: hidden !important;
  border-radius: 12px !important;
  mix-blend-mode: var(--p-link-tile-product-mix-blend-mode) !important;
}
::slotted(a) {
  position: absolute !important;
  border-radius: 8px !important;
  top: 0 !important;
  left: 0 !important;
  right: 0 !important;
  bottom: 0 !important;
  z-index: 1 !important;
  text-indent: -999999px !important;
}
::slotted(a:focus) {
  outline: 2px solid #1A44EA !important;
  outline-offset: 2px !important;
}
::slotted(a:focus:not(:focus-visible)) {
  outline-color: transparent !important;
}
.link-overlay {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  border-radius: 8px;
}
.link-overlay:focus {
  outline: 2px solid #1A44EA;
  outline-offset: 2px;
}
.link-overlay:focus:not(:focus-visible) {
  outline-color: transparent;
}
.root {
  display: flex;
  flex-direction: column;
  aspect-ratio: 3 / 4;
  overflow: hidden;
  box-sizing: border-box;
  border-radius: 8px;
  padding: clamp(8px, 0.5vw + 6px, 16px);
  color: #010205;
  background-color: #EEEFF2;
}
.header {
  display: flex;
  justify-content: space-between;
  font-size: clamp(0.81rem, 0.23vw + 0.77rem, 0.88rem);
}
.header slot[name="header"] {
  display: block;
}
.like-button {
  height: fit-content;
  position: relative;
  z-index: 2;
}
.text-container {
  display: flex;
  justify-content: center;
  flex-direction: column;
  margin: auto;
}
.text {
  text-align: center;
  margin: 0;
}
.text__heading {
  font: normal normal 600 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  padding-bottom: 2px;
  min-height: calc(calc(6px + 2.125ex) * 2);
  display: flex;
  justify-content: center;
  align-items: flex-end;
}
.text__price, .text__info {
  font-weight: 400;
}
.text__price {
  font: normal normal 400 clamp(0.81rem, 0.23vw + 0.77rem, 0.88rem)/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
}
.text__info {
  font: normal normal 400 .75rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  color: #535457;
}
.image-container {
  margin: clamp(8px, 0.5vw + 6px, 16px) auto clamp(4px, 0.25vw + 3px, 8px) auto;
  padding: 0;
  overflow: hidden;
  aspect-ratio: 8 / 9;
  transition: transform var(--p-transition-duration, 0.4s) cubic-bezier(0.25,0.1,0.25,1);
  max-height: 100%;
}
@media(min-width:760px) {
  .image-container {
    padding: 0 clamp(16px, 1.25vw + 12px, 36px);
  }
}

@media(hover:hover) {
  .root:hover .image-container {
    transform: scale3d(1.05,1.05,1.05);
  }
  .like-button:hover {
    cursor: pointer;
  }
}
"
`;
